packages/math/src/pool/concentrated/quotes.ts

```diff
@@ -0,0 +1,215 @@
+  amountRemaining: Dec;
+  amountCalculated: Dec;
+  inittedTickIndex: number;
+  sqrtPrice: Dec;
+  liquidity: Dec;
+  feeGrowthGlobal: Dec;
+}
+
+function calcOutGivenIn(
+  tokenIn: Coin,
+  tokenDenom0: string,
+  poolLiquidity: Dec,
+  inittedTicksWithNetLiquidity: TickWithNetLiquidity[],
+  curTick: Int,
+  curSqrtPrice: Dec,
+  precisionFactorAtPriceOne: number,
+  swapFee: Dec,
+  priceLimit = new Dec(0)
+): Int {
+  const isZeroForOne = tokenIn.denom === tokenDenom0;
+  if (isZeroForOne && priceLimit.equals(new Dec(0))) {
+    priceLimit = minSpotPrice;
+  } else if (!isZeroForOne && priceLimit.equals(new Dec(0))) {
+    priceLimit = maxSpotPrice;
+  }
+
+  const sqrtPriceLimit = approxSqrt(priceLimit);
+  const swapStrategy = makeSwapStrategy(isZeroForOne, sqrtPriceLimit, swapFee);
+  const tokenInAmountSpecified = new Dec(tokenIn.amount);
+
+  const curTickWithNetLiq = inittedTicksWithNetLiquidity.find(({ tickIndex }) =>
+    tickIndex.equals(curTick)
+  );
+  if (!curTickWithNetLiq) {
+    throw new Error("curTickNet not found in inittedTicksWithNetLiquidity");
+  }
+  const curTickArrIndex =
+    inittedTicksWithNetLiquidity.indexOf(curTickWithNetLiq);
+  if (curTickArrIndex < 0) {
+    throw new Error(
+      "curTickWithNetLiq not found in inittedTicksWithNetLiquidity"
+    );
+  }
+
+  const swapState: SwapState = {
+    amountRemaining: tokenInAmountSpecified, // tokenIn
+    amountCalculated: new Dec(0), // tokenOut
+    inittedTickIndex: swapStrategy.initTickValue(curTickArrIndex),
+    sqrtPrice: curSqrtPrice,
+    liquidity: poolLiquidity,
+    feeGrowthGlobal: new Dec(0),
+  };
+
+  while (
+    swapState.amountRemaining.gt(smallestDec) &&
+    !swapState.sqrtPrice.equals(sqrtPriceLimit)
+  ) {
+    const nextTick: TickWithNetLiquidity | undefined =
+      inittedTicksWithNetLiquidity?.[swapState.inittedTickIndex];
+    if (!nextTick) {
+      throw new TickOverflowError("Not enough ticks to calculate swap");
+    }
+
+    const nextTickSqrtPrice = tickToSqrtPrice(
+      nextTick.tickIndex,
+      precisionFactorAtPriceOne
+    );
+
+    const sqrtPriceTarget = swapStrategy.getSqrtTargetPrice(nextTickSqrtPrice);
+
+    const {
+      sqrtPriceNext,
+      amountInConsumed: amountOneIn,
+      amountOutComputed: amountZeroOut,
+      feeChargeTotal,
+    } = swapStrategy.computeSwapStepOutGivenIn(
+      swapState.sqrtPrice,
+      sqrtPriceTarget,
+      swapState.liquidity,
+      swapState.amountRemaining
+    );
+
+    swapState.sqrtPrice = sqrtPriceNext;
+    swapState.amountRemaining = swapState.amountRemaining.sub(
+      amountOneIn.add(feeChargeTotal)
+    );
+    swapState.amountCalculated = swapState.amountCalculated.add(amountZeroOut);
+
+    if (nextTickSqrtPrice.equals(sqrtPriceNext)) {
+      const liquidityNet = swapStrategy.setLiquidityDeltaSign(
+        new Dec(nextTick.netLiquidity.toString())
+      );
+
+      swapState.liquidity = addLiquidity(swapState.liquidity, liquidityNet);
+    }
```

Comment:
Is that initting a new tick, or creating a new tickIndex? Do we insert some value into the `inittedTicksWithNetLiquidity` array? Not sure how to handle the update to `inittedTickIndex`, which is handled differently on the FE since we can't query or init ticks on our end. May still be missing knowledge of how we handle stopping an iteration mid-tick.

Perhaps we take our PriceToTick computed value and find() it in `inittedTicksWithNetLiquidity` array?

Reply:
Upon thinking more about how FE differs from chain, I suspect we might not need to port it here. But let me try breaking down the logic.

That branch is executed when a swap does not reach the next initialized tick.

As a result, we stop mid-way between one initialized tick and the next initialized tick.

At that point, we only have the knowledge of the cur sqrt price. Therefore, we need to compute the cur tick from the sqrt price. The tick itself is not initialized.

Once we have the mid-non-initialized tick, we compute the next initialized tick at the next iteration of the loop:
https://github.com/osmosis-labs/osmosis/blob/d61c0eda66f7c94110307e2bca97a461ee809afd/x/concentrated-liquidity/swaps.go#L339

On-chain, we compute the next tick by reading it from the state. Here, I think we should be able to just move forward to the next element in the list returned by the chain query.